### PR TITLE
[publication] 21.0 release publication email verify fix

### PR DESCRIPTION
**Pull request by maltheism**
### Brief summary of changes

Incorporated the frontend browser check for validation of email by changing type to email.

### This resolves issue...

- [ ] Redmine? #

- [x] Github? #4704

### To test this change...

- [x] Test the publication module with an email containing the '+' symbol.

### CODE REVIEW

File: modules/publication/jsx/uploadForm.js [hunk 1]
```diff
@@ -27,12 +27,8 @@ class PublicationUploadForm extends React.Component {
...
   validateEmail(field, email) {
     let formErrors = this.state.formErrors;
-
-    // don't supply error if email is blank
-    if (email === '' || email === null || email === undefined) {
-      delete formErrors[field];
-      // if email is invalid, set error, else nullify error
-    } else if (/^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/.test(email) === false) {
+    const emailDiv = document.getElementById(field);
+    if (!emailDiv.checkValidity()) {
```
Comment (driusan) [inline modules/publication/jsx/uploadForm.js]: this doesn't appear to be defined?

File: jsx/Form.js [hunk 2]
```diff
@@ -1365,6 +1365,57 @@ StaticElement.defaultProps = {
...
+ * Email element component.
+ * Used to displays plain/formatted email input as part of a form
+ *
+ * To pass a formatted text, you need to wrap it in a single parent element.
+ * Example usage:
+ *
+ * ```
+ * let myText = (<span>This is my <b>text</b></span>);
+ * <EmailElement
+ *    text={myText}
```
Comment (driusan) [inline jsx/Form.js]: why not use this.props.children? That would make this cleaner, I think

File: modules/publication/jsx/viewProject.js [hunk 3]
```diff
@@ -257,10 +247,11 @@ class ViewProject extends React.Component {
           label="Lead Investigator"
           text={this.state.formData.leadInvestigator}
         />
-        <StaticElement
-         name="leadInvestigatorEmail"
-         label="Lead Investigator Email"
-         text={this.state.formData.leadInvestigatorEmail}
+        <EmailElement
+          id={'leadInvestigatorEmail'}
+          name="leadInvestigatorEmail"
```
Comment (zaliqarosli) [inline modules/publication/jsx/viewProject.js]: please keep consistent with double/single quotes and i don't think the value of id needs curly brackets too

File: jsx/Form.js [hunk 4]
```diff
@@ -1365,6 +1365,57 @@ StaticElement.defaultProps = {
...
+    super(props);
+  }
+  render() {
+    return (
+      <div className="row form-group">
+        <label className="col-sm-3 control-label">
+          {this.props.label}
+        </label>
+        <div className="col-sm-9">
+          <input className="form-control-static" type={'email'} id={this.props.id}>
```
Comment (zaliqarosli) [inline jsx/Form.js]: i think just 'email' without the curlies is sufficient no?

File: jsx/Form.js [hunk 5]
```diff
@@ -1365,6 +1365,57 @@ StaticElement.defaultProps = {
...
+            {this.props.text}
+          </input>
+        </div>
+      </div>
+    );
+  }
+}
+
+EmailElement.propTypes = {
+  label: PropTypes.string,
```
Comment (zaliqarosli) [inline jsx/Form.js]: you can add name prop

File: modules/publication/jsx/projectFields.js [hunk 6]
```diff
@@ -39,7 +39,7 @@ class EmailElement extends React.Component {
         </label>
         <div className="col-sm-7">
           <input
-            type="text"
+            type="email"
```
Comment (zaliqarosli) [inline modules/publication/jsx/projectFields.js]: i'm a bit confused here. do you also want to remove this EmailElement component out of this jsx file?

File: jsx/Form.js [hunk 7]
```diff
@@ -1365,6 +1365,57 @@ StaticElement.defaultProps = {
...
 
+/**
+ * Email element component.
+ * Used to displays plain/formatted email input as part of a form
+ *
+ * To pass a formatted text, you need to wrap it in a single parent element.
+ * Example usage:
+ *
+ * ```
+ * let myText = (<span>This is my <b>text</b></span>);
```
Comment (zaliqarosli) [inline jsx/Form.js]: im curious .. whats the reason behind allowing formatted text? and what is this text attribute? is it just the value of the field? i am wondering why it's not going into a 'value' attribute like the others

File: jsx/Form.js (hunk 4 above)
Comment (maltheism) [reply jsx/Form.js]: @zaliqarosli My IDE does that by default, I don't think it matters but removed.

File: modules/publication/jsx/projectFields.js (hunk 6 above)
Comment (maltheism) [reply modules/publication/jsx/projectFields.js]: @zaliqarosli I think I left this because the component has different column than the other component.

File: modules/publication/jsx/uploadForm.js (hunk 1 above)
Comment (maltheism) [reply modules/publication/jsx/uploadForm.js]: @driusan I think I modified the previous code to be similar but the element emailDiv should be defined from the 'field' parameter that's passed into validateEmail.

Comment (driusan) [reply modules/publication/jsx/uploadForm.js]: I mean the function `checkValidity` isn't defined.

File: modules/publication/jsx/viewProject.js [hunk 8]
```diff
@@ -257,10 +247,11 @@ class ViewProject extends React.Component {
           label="Lead Investigator"
           text={this.state.formData.leadInvestigator}
         />
-        <StaticElement
-         name="leadInvestigatorEmail"
-         label="Lead Investigator Email"
-         text={this.state.formData.leadInvestigatorEmail}
+        <EmailElement
```
Comment (zaliqarosli) [inline modules/publication/jsx/viewProject.js]: it looks to me that this EmailElement is the one from jsx/Form.js which is an input element `<input type='email'/>` except this use of it is replacing a StaticElement (a display) and not an input element?

File: jsx/Form.js [hunk 9]
```diff
@@ -1365,6 +1365,48 @@ StaticElement.defaultProps = {
   text: null,
 };
 
+/**
+ * Email element component.
+ * Used to displays plain/formatted email input as part of a form.
```
[the remark this answers is not in the document]
Comment (maltheism) [reply jsx/Form.js]: @zaliqarosli I'll redefine the comment here. Also to answer your other comment. I left the module alone without using this component that was made with this PR because of the css is different than a typical form. I didn't feel this PR was targeting that but just the bug of solving correct email verification.

File: modules/publication/jsx/viewProject.js (hunk 8 above)
Comment (maltheism) [reply modules/publication/jsx/viewProject.js]: @zaliqarosli I'm confused how you phrased this question. The reason for the EmailElement is similar to why there was a request for a unique password component in the react login PR.

File: modules/publication/jsx/uploadForm.js (hunk 1 above)
Comment (maltheism) [reply modules/publication/jsx/uploadForm.js]: @driusan See https://developer.mozilla.org/en-US/docs/Web/API/HTMLSelectElement/checkValidity

File: modules/publication/jsx/viewProject.js (hunk 8 above)
Comment (zaliqarosli) [reply modules/publication/jsx/viewProject.js]: is the password component you're mentioning an input element or a display element?

File: jsx/Form.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (zaliqarosli) [reply jsx/Form.js]: > I left the module alone without using this component that was made with this PR because of the css is different than a typical form.

I don't understand why a component is made in this PR but then is not used

[the remark this answers is not in the document]
Comment (zaliqarosli) [reply jsx/Form.js]: perhaps this PR includes more code changes than is needed to target the actual bugfix? I'm not quite sure what I'm reviewing here. it looks to me that there are 2 different EmailElement components with the same name, i'm unsure which one is being used, and when it is being used, its to replace a StaticElement (a display) and not an input element 😕

File: modules/publication/jsx/viewProject.js (hunk 8 above)
Comment (maltheism) [reply modules/publication/jsx/viewProject.js]: Hey @zaliqarosli, the PR for upgrading the Login module to react is what i'm referring to and when I'm referencing the new password component in that PR (you suggested it). Instead of modifying one of the components to add a property that would change the text to be type "password" it was suggested to just make another component. I thought similar for creating an email component because of the type being email.

Comment (zaliqarosli) [reply modules/publication/jsx/viewProject.js]: okay great. I agree that we should be making separate components. Can you point me to where in the code in this PR that you are using this new EmailComponent that you added to jsx/Form.js?

Comment (maltheism) [reply modules/publication/jsx/viewProject.js]: I'll delete the component then. As your above comment mentions it confuses the review process.

File: jsx/Form.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (maltheism) [reply jsx/Form.js]: Hi @zaliqarosli, I'll delete the component then. As your comment mentions it confuses the review process.

File: modules/publication/jsx/viewProject.js (hunk 8 above)
Comment (zaliqarosli) [reply modules/publication/jsx/viewProject.js]: no need to delete it. if it's not being used in this PR, then I think it's worth it for you to open up a separate PR for this new EmailElement component and include a use case there

Comment (maltheism) [reply modules/publication/jsx/viewProject.js]: Hi @zaliqarosli, I removed the component from this PR.

Comment (zaliqarosli) [inline modules/publication/jsx/viewProject.js]: this StaticElement should stay a static element right?

File: modules/publication/jsx/projectFields.js [hunk 10]
```diff
@@ -403,11 +411,15 @@ class ProjectFormFields extends React.Component {
...
-        <StaticElement
-          text={voiHelp}
-        />
+        <div className="row form-group">
+          <label className="col-sm-3 control-label">
+          </label>
+          <div className="col-sm-9">
+            <div className="form-control-static">{voiHelp}</div>
+          </div>
+        </div>
```
Comment (zaliqarosli) [inline modules/publication/jsx/projectFields.js]: what's the reason for this change? it looks to me that the html here is the same as is in the StaticElement render function

File: modules/publication/jsx/uploadForm.js [hunk 11]
```diff
@@ -195,14 +191,18 @@ class PublicationUploadForm extends React.Component {
...
+            Propose a new project
+          </h3>,
+          <TextboxElement
+            name="title"
+            label="Title"
+            onUserInput={this.setFormData}
+            required={true}
+            value={this.state.formData.title}
+          />,
+        </div>,
```
Comment (zaliqarosli) [inline modules/publication/jsx/uploadForm.js]: createElements is an array of jsx. i don't think you need the div that you added here and it's also incorrect syntax right? you're nesting elements of an array within div tags.

File: modules/publication/jsx/viewProject.js [hunk 12]
```diff
@@ -25,16 +25,6 @@ class ViewProject extends React.Component {
 
   validateEmail(field, email) {
     let formErrors = this.state.formErrors;
-
-    // don't supply error if email is blank
```
Comment (zaliqarosli) [inline modules/publication/jsx/viewProject.js]: what has this validation been replaced with?

File: modules/publication/jsx/uploadForm.js (hunk 11 above)
Comment (maltheism) [reply modules/publication/jsx/uploadForm.js]: @zaliqarosli The extra div is fixing the error "Warning: Each child in an array or iterator should have a unique "key" prop." from displaying in the console. I'll change the formatting to be createElement = () instead of an array. Currently we don't have a way to supply a key to TextboxElement or I would find the extra div unnecessary

File: modules/publication/jsx/projectFields.js (hunk 10 above)
Comment (maltheism) [reply modules/publication/jsx/projectFields.js]: @zaliqarosli There is a slight change if you compare the render of StaticElement to my change. Turns out this change solved the console error: "Warning: validateDOMNesting(...): <div> cannot appear as a descendant of <p>." but maybe I should modify the StaticElement and get rid of the use of <p>. What do you think?

Comment (zaliqarosli) [reply modules/publication/jsx/projectFields.js]: could you try changing to <div> tags in `voiHelp` to <span>? I think this would be the best solution

File: modules/publication/jsx/viewProject.js (hunk 12 above)
Comment (zaliqarosli) [reply modules/publication/jsx/viewProject.js]: hey @maltheism, what's your solution for this? the validation is still missing

File: modules/publication/jsx/projectFields.js (hunk 10 above)
Comment (maltheism) [reply modules/publication/jsx/projectFields.js]: @zaliqarosli sure why not

File: modules/publication/jsx/viewProject.js (hunk 12 above)
Comment (maltheism) [reply modules/publication/jsx/viewProject.js]: Hey @zaliqarosli I'm not sure why the validateEmail is showing here on GitHub. I've removed the validateEmail function here if you checkout the PR and look at the code. The validation for the email is being done in the uploadForm.js

Comment (zaliqarosli) [reply modules/publication/jsx/viewProject.js]: I know that you've removed the validateEmail function here. my question is why? shouldn't we keep the validation here?

it looks to me that validateEmail prop is to be supplied to the `<ProjectFormFields />` component of which there are two - one in `viewProject.js` and one in `uploadForm.js`. it looks to me that validateEmail function needs to be defined in both` viewProject.js` and `uploadForm.js`.

You are removing the validateEmail function from `viewProject.js`. If it looks to you that `viewProject.js` doesn't need the validateEmail function, please explain to me why. And explain to me why the `<ProjectFormFields />` component in this same class doesn't need the validateEmail prop (which you are also removing)

Comment (maltheism) [reply modules/publication/jsx/viewProject.js]: @zaliqarosli Thanks for all the help!

**Comment by maltheism** [comment]
Updated PR to just use type=email for the validation because the browser handles it when type is email.

File: modules/publication/jsx/viewProject.js [hunk 13]
```diff
@@ -258,9 +242,10 @@ class ViewProject extends React.Component {
           text={this.state.formData.leadInvestigator}
         />
         <StaticElement
-         name="leadInvestigatorEmail"
-         label="Lead Investigator Email"
-         text={this.state.formData.leadInvestigatorEmail}
+          id='leadInvestigatorEmail'
```
Comment (zaliqarosli) [inline modules/publication/jsx/viewProject.js]: i think you can get ride of this id too

Comment (maltheism) [reply modules/publication/jsx/viewProject.js]: thanks @zaliqarosli it should be fixed now.
